src/encode_data.circom

```diff
@@ -0,0 +1,271 @@
+    encodeSpotTradeNewOrder2AmountSell.in <== newOrder2AmountSell;
+    encodeSpotTradeNewOrder2AmountSell.enabled <== 1;
+    for (var i = 0; i < floats; i++) {
+        encodedSpotTradeTx[i+offset] <== useSpotTrade*encodeSpotTradeNewOrder2AmountSell.out[i];
+    }
+    offset += floats;
+    component encodeSpotTradeNewOrder2AmountBuy = Num2BitsIfEnabled(floats);
+    encodeSpotTradeNewOrder2AmountBuy.in <== newOrder2AmountBuy;
+    encodeSpotTradeNewOrder2AmountBuy.enabled <== 1;
+    for (var i = 0; i < floats; i++) {
+        encodedSpotTradeTx[i+offset] <== useSpotTrade*encodeSpotTradeNewOrder2AmountBuy.out[i];
+    }
+    offset += floats;
+    component encodeSpotTradeNewOrder2ID = Num2BitsIfEnabled(orderLevels);
+    encodeSpotTradeNewOrder2ID.in <== newOrder2ID;
+    encodeSpotTradeNewOrder2ID.enabled <== 0;
+    for (var i = 0; i < orderLevels; i++) {
+        encodedSpotTradeTx[i+offset] <== useSpotTrade*encodeSpotTradeNewOrder2ID.out[i];
+    }
+    offset += orderLevels;
+    //filling reset part by 0
+    assert(offset <= encodeLength);
+    for (var i = 0; i < encodeLength - offset; i++) {
+        encodedSpotTradeTx[i+offset] <== 0;
+    }
+
+    //start scheme {l2Key} encoding
+    signal useL2Key;
+    signal encodedL2KeyTx[encodeLength];
+    
+    useL2Key <== isL2KeyUpdated;        
+    //this constraints ensure l2key can only be updated under a 'dummy' deposit tx
+    signal notDepositFlag;
+    notDepositFlag <== isWithdraw + isTransfer + isSpotTrade;
+    isL2KeyUpdated * notDepositFlag === 0;
+    //(isWithdraw + isTransfer + isSpotTrade) * isL2KeyUpdated === 0;
+    amount * isL2KeyUpdated === 0;
+    encodedL2KeyTx[0] <== useL2Key;
+    encodedL2KeyTx[1] <== 0;
+    encodedL2KeyTx[2] <== 0;
+    useL2Key*(useL2Key - 1) === 0;
+    schemeCheck += useL2Key;
+    offset = 3;
+    //start filling encoded part
+    component encodeL2KeyAccountID1 = Num2BitsIfEnabled(accountLevels);
+    encodeL2KeyAccountID1.in <== accountID1;
+    encodeL2KeyAccountID1.enabled <== 1;
+    for (var i = 0; i < accountLevels; i++) {
+        encodedL2KeyTx[i+offset] <== useL2Key*encodeL2KeyAccountID1.out[i];
+    }
+    offset += accountLevels;
+    component encodeL2KeySign2 = Num2BitsIfEnabled(1);
+    encodeL2KeySign2.in <== sign2;
+    encodeL2KeySign2.enabled <== 1;
+    for (var i = 0; i < 1; i++) {
+        encodedL2KeyTx[i+offset] <== useL2Key*encodeL2KeySign2.out[i];
+    }
+    offset += 1;
+    component encodeL2KeyAy2 = Num2BitsIfEnabled(254);
+    encodeL2KeyAy2.in <== ay2;
+    encodeL2KeyAy2.enabled <== 1;
+    for (var i = 0; i < 254; i++) {
+        encodedL2KeyTx[i+offset] <== useL2Key*encodeL2KeyAy2.out[i];
+    }
+    offset += 254;
+    //filling reset part by 0
+    assert(offset <= encodeLength);
+    for (var i = 0; i < encodeLength - offset; i++) {
+        encodedL2KeyTx[i+offset] <== 0;
+    }
+
+
+    signal finalCheck;
+    finalCheck <== schemeCheck;
+    finalCheck * (finalCheck - 1) === 0;
```

Comment:
I think it should be `finalCheck === 1`? so that scheme is one of "useCommon", "useSpotTrade" or "useL2Key".

Reply:
ok.... so we also have another type `useNop`